<launch>
  <arg name="mav_name" default="firefly"/>
  <arg name="enable_logging" default="false" />
  <arg name="enable_ground_truth" default="true" />
  <arg name="log_file" default="$(arg mav_name)" />
  <!-- <arg name="world_name" default="rectangle_tag"/> -->
  <!-- <arg name="world_name" default="compare"/>  -->
  <arg name="world_name" default="corritags"/>
  <!-- <arg name="world_name" default="corri"/> -->
  <!-- <arg name="world_name" default="home1"/> -->
  <!-- <arg name="world_name" default="basic"/> -->
  <!-- <arg name="world_name" default="t"/> -->
  <arg name="debug" default="false"/>
  <arg name="gui" default="true"/>
  <arg name="paused" default="true"/>
  <arg name="verbose" default="false"/>


  <arg name="x" default="2"/>
  <arg name="y" default="0.0"/>
  <arg name="z" default="0.0"/>

	<!-- ***************1.quadrotor****************-->
	<include file="$(find gazebo_ros)/launch/empty_world.launch">
		<arg name="world_name" value="$(find dvins_core)/worlds/$(arg world_name).world" />
		<arg name="debug" value="$(arg debug)" />
		<arg name="paused" value="$(arg paused)" />
		<arg name="gui" value="$(arg gui)" />
		<arg name="verbose" value="$(arg verbose)"/>
	</include>

	<group ns="$(arg mav_name)">
		<rosparam param="wp_x">0</rosparam>
		<rosparam param="wp_y">1.0</rosparam>
		<rosparam param="wp_z">3.0</rosparam>
		<include file="$(find dvins_core)/launch/spawn_mav.launch">
			<arg name="mav_name" value="$(arg mav_name)" />
			<!-- <arg name="model" value="$(find dvins_core)/urdf/firefly_generic_odometry_mavcablepayload.gazebo" /> -->
			<arg name="model" value="$(find dvins_core)/urdf/firefly_generic_odometry_mavcablepayload_camera.gazebo" />
			<arg name="enable_logging" value="$(arg enable_logging)" />
			<arg name="enable_ground_truth" value="$(arg enable_ground_truth)" />
			<arg name="log_file" value="$(arg mav_name)1"/>
			<arg name="z" value="0.07"/>
		</include>
	</group>
	<group ns="payload">
		<include file="$(find dvins_core)/launch/spawn_payload.launch">
		<arg name="mav_name1" value="firefly" />
		<arg name="payload_mass" value="0.1" />
		<arg name="body_width" value="0.1" />
		<arg name="body_length" value="0.1" />
		<arg name="body_height" value="0.1" />
		<arg name="z" value="0.051" />
		</include>
	</group>
	
	<!-- ***************2.little car****************-->
	<include file="$(find dvins_eval)/launch/car.launch">
		<arg name="higher_level" value="false"/>
	</include>
	<!-- ***************3.detection****************-->
	<include file="$(find dvins_core)/launch/tag.launch">
    <arg name="show_detections" value="false" />
		<arg name="camera" value="/firefly/camera_2/image_raw"/>
		<arg name="camera_info" value="/firefly/camera_2/camera_info"/>
	</include>

	<!-- ***************4.estimator and controller****************-->
	<group ns="$(arg mav_name)">
		<node name="dvins_estimator" pkg="dvins_core" type="dvins_estimator2" output="screen"/>
		<!-- <node name="dvins_estimator" pkg="dvins_core" type="dvins_estimator" output="screen"/> -->
		<!-- <node name="dvins_estimator" pkg="dvins_core" type="dvins_estimator1" output="screen"/> -->

		<!-- copied lpcn -->
		<!-- <node name="lpcn" pkg="dvins_core" type="lpcn" output="screen">
			<rosparam command="load" file="$(find dvins_core)/resource/lee_controller_$(arg mav_name).yaml" />
			<rosparam command="load" file="$(find dvins_core)/resource/$(arg mav_name).yaml" />
			<remap from="odometry" to="odometry_sensor1/odometry" />
			<remap from="payload_odometry" to="/payload/odometry_sensor2/odometry" />
		</node> -->

		<!-- copied lcn -->
		<!-- <node name="lee_position_controller_node" pkg="dvins_core" type="lee_position_controller_node" output="screen">
			<rosparam command="load" file="$(find dvins_core)/resource/lee_controller_$(arg mav_name).yaml" />
			<rosparam command="load" file="$(find dvins_core)/resource/$(arg mav_name).yaml" />
			<remap from="odometry" to="odometry_sensor1/odometry" />
			<remap from="payload_odometry" to="/payload/odometry_sensor2/odometry" />
		</node> -->

		<!-- origin lee_payload -->
		<!-- <node name="lee_payload" pkg="dvins_core" type="lee_payload" output="screen">
			<rosparam command="load" file="$(find dvins_core)/resource/lee_controller_$(arg mav_name).yaml" />
			<rosparam command="load" file="$(find dvins_core)/resource/$(arg mav_name).yaml" />
			<remap from="odometry" to="odometry_sensor1/odometry" />
			<remap from="payload_odometry" to="/payload/odometry_sensor2/odometry" />
		</node> -->
	
		<!-- <node name="hover_loop" pkg="dvins_core" type="hover_loop" output="screen"/> -->

		<node name="hovering_example" pkg="dvins_core" type="hovering_example" output="screen"/>
		<node name="trajectory_publisher_rectangle" pkg="dvins_core" type="trajectory_publisher_rectangle" />
	</group>
</launch>
